File: 08.py
import fileinput

def run(program):
    ip = acc = 0
    seen = set()
    while ip < len(program):
        if ip in seen:
            return acc, False
        seen.add(ip)
        op, arg = program[ip]
        if op == 'acc':
            acc += int(arg)
        if op == 'jmp':
            ip += int(arg) - 1
        ip += 1
    return acc, True

def part1(program):
    return run(program)[0]

def part2(program):
    m = {'jmp': 'nop', 'nop': 'jmp'}
    for i in range(len(program)):
        op, arg = program[i]
        p = list(program)
        p[i] = (m.get(op, op), arg)
        acc, ok = run(p)
        if ok:
            return acc

program = [x.split() for x in fileinput.input()]
print(part1(program))
print(part2(program))

File: 08.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

type Instruction struct {
	Op  string
	Arg int
}

type Program []Instruction

func parseProgram(r io.Reader) Program {
	var program Program
	scanner := bufio.NewScanner(r)
	for scanner.Scan() {
		fields := strings.Fields(scanner.Text())
		arg, _ := strconv.ParseInt(fields[1], 0, 0)
		program = append(program, Instruction{fields[0], int(arg)})
	}
	return program
}

func runProgram(program Program) (int, bool) {
	var ip, acc int
	seen := make(map[int]bool)
	for ip < len(program) {
		if seen[ip] {
			return acc, false
		}
		seen[ip] = true
		op := program[ip].Op
		arg := program[ip].Arg
		switch op {
		case "acc":
			acc += arg
		case "jmp":
			ip += arg - 1
		}
		ip++
	}
	return acc, true
}

func part1(program Program) int {
	acc, _ := runProgram(program)
	return acc
}

func part2(program Program) int {
	for i := 0; i < len(program); i++ {
		instruction := program[i]
		switch instruction.Op {
		case "jmp":
			program[i].Op = "nop"
		case "nop":
			program[i].Op = "jmp"
		}
		acc, ok := runProgram(program)
		program[i] = instruction
		if ok {
			return acc
		}
	}
	return 0
}

func main() {
	program := parseProgram(os.Stdin)
	fmt.Println(part1(program))
	fmt.Println(part2(program))
}
